model-optimizer/extensions/front/tf/CTCGreedyDecoderReplacement.py

```diff
@@ -14,67 +14,223 @@
+from extensions.front.Pack import Pack
+from extensions.front.FillToBroadcast import FillToBroadcast
+from mo.front.common.partial_infer.utils import int64_array
 from mo.front.common.replacement import FrontReplacementSubgraph
-from mo.graph.graph import Graph
-from mo.ops.const import Const
-from mo.utils.error import Error
+from mo.front.tf.graph_utils import create_op_with_const_inputs
+from mo.graph.graph import Graph, rename_nodes
+from mo.ops.broadcast import Broadcast
+from mo.ops.concat import Concat
+from mo.ops.squeeze import Squeeze
+from mo.ops.unsqueeze import Unsqueeze
 
 
 class CTCGreedyDecoderReplacement(FrontReplacementSubgraph):
+    """
+    TensorFlow CTCGreedyDecoder produces output in a sparse tensor that is not supported by Inference Engine and
+    Inference Engine's CTCGreedyDecoder has different output that is in a dense format. So this transformation
+    intents to replace TF CTCGreedyDecoder+SparseToDense with IE one.
+    Also Inference Engine's CTCGreedyDecoder has a specific format for the second input tensor, a sequence length,
+    different from TF's one so this transformation cares about transformation of its format.
+    The second input to the CTCGreedyDecoder in the TensorFlow is a 1D tensor with sequence lengths. In the Inference
+    Engine the second input to the CTCGreedyDecoder is a 2D tensor, a sequence mask, where the first element
+    in each row is equal to 1 and all others in the tail are equal to 0. The number of ones represents
+    a sequence length.
+    """
+    enabled = True
+
+    def run_after(self):
+        # CTCGreedyDecoderReplacement is not reshape-able transformation
+        # so reshape-able CTCGreedyDecoderReplacement2 transformation is applied first
+        return [CTCGreedyDecoderReplacement2]
+
+    @staticmethod
+    def pattern(**kwargs):
+        return dict(
+            nodes=[('decoder', dict(op='CTCGreedyDecoder')),
+                   ('cast', dict(op='Cast')),
+                   ('sparse_to_dense', dict(op='SparseToDense'))
+                   ],
+            edges=[('decoder', 'sparse_to_dense', {'out': 0}),
+                   ('decoder', 'cast', {'out': 1}),
+                   ('cast', 'sparse_to_dense', {'out': 0})
+                   ]
+        )
+
+    def replace_sub_graph(self, graph: Graph, match: dict):
+        # TODO: Once Inference Engine's CTCGreedyDecoder starts to support sequence length format like in TensorFlow,
+        # CTCGreedyDecoderReplacement2 needs to be removed and CTCGreedyDecoderReplacement, a more generic
+        # transformation, needs to be adopted for all cases
+        ctc_greedy_decoder = match['decoder']
+        cast = match['cast']
+        sparse_to_dense = match['sparse_to_dense']
+        sparse_to_dense_name = sparse_to_dense.soft_get('name', sparse_to_dense.id)
+
+        # remove edges from CTCGreedyDecoder node since top nodes willbe changed
+        graph.remove_edge(ctc_greedy_decoder.id, sparse_to_dense.id)
```

Comment:
Can we replace these function calls with methods like `node.in_port(XXX).disconnect()`?

Reply:
Yes, fixed.